Added one-time permissions article (#6948)
GoogleChrome/developer.chrome.com: site/en/blog/one-time-permissions/index.md

---
layout: 'layouts/blog-post.njk'
title: 'One-time permissions in Chrome'
description: >
  "Allow this time" is a one-time option for permission grants in Chrome. It will initially be available on desktop only for geolocation, camera, and microphone.
authors:
  - maudn
date: 2023-07-31
updated: 2023-07-31
hero: image/O2RNUyVSLubjvENAT3e7JSdqSOx1/FRytD8YVBZi9P3hAcOXv.jpeg

> alt: Permission prompt in Chrome, with the Allow this time option

Reviewer: ```suggestion
alt: Permission prompt in Chrome, with the Allow this time option.
```

tags:
  - capabilities
  - privacy
---

> With a gradual rollout from [Chrome 116](https://chromiumdash.appspot.com/schedule), we will be adding the **Allow this time** option to permission prompts. Our goal is to make it easier for people to use powerful web capabilities on their own terms. **Allow this time** will initially be available on desktop for some of the most common permissions: **geolocation**, **camera**, and **microphone**. Permission prompts for other capabilities and on the mobile web are unchanged for now, but may change in the future.

Reviewer: ```suggestion
With a gradual rollout from [Chrome 116](https://chromiumdash.appspot.com/schedule), we will be adding the **Allow this time** option to permission prompts. Our goal is to make it easier for people to use powerful web capabilities on their own terms. **Allow this time** will initially be available on desktop for some of the most common permissions: geolocation, camera, and microphone. Permission prompts for other capabilities and on the mobile web are unchanged for now, but may change in the future.
```

**If you're a site owner, review [Impact and Recommendations](#impact-and-recommendations), and try the [demo](#demo).**

## A new UX for permissions

### Permissions UX before Chrome 116

<figure>

> {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/TcL9YukUHlkdn1bLbZTv.png", alt="Permission prompt until now", width="800", height="427" %}

Reviewer: ```suggestion
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/TcL9YukUHlkdn1bLbZTv.png", alt="Permission prompt showing only block and allow.", width="800", height="427" %}
```

  <figcaption>

>     Permission prompt until now

Reviewer: ```suggestion
    Permission prompt before Chrome 116.
```

  </figcaption>
</figure>

Users are presented with three options:

- **Allow**: Persistent allow.
- **Block**: Persistent block.
- Clicking the **x** button: Temporary block. The site can prompt the user again later, at most three times.

Permissions can be managed via the **Site controls** in the address bar or via the **Site settings**.

{% Aside 'key-term' %}
*Site controls*: User-facing information and controls for a site, accessible after clicking the icon in the address bar.

<figure>

> {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/UD9Don3Ks7ZYL0cF2c2c.png", alt="Site controls", width="483", height="370" %}

Reviewer: ```suggestion
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/UD9Don3Ks7ZYL0cF2c2c.png", alt="The site controls menu in Chrome.", width="483", height="370" %}
```

  <figcaption>

>     Site controls

Reviewer: ```suggestion
    Site controls in Chrome.
```

  </figcaption>
</figure>
{% endAside %}

{% Aside 'key-term' %}
*Site settings*: User-facing information and controls for a site. Site settings can be accessed by clicking **Site settings** in the **Site controls** menu, or via Chrome's **Settings** page (`chrome://settings` or the **Settings** entry in the [three-dot-menu](https://support.google.com/chrome/answer/114662?hl=en)).
{% endAside %}

### New permissions UX from Chrome 116

<figure>

> {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/7aSSSclHdEj33k5ifeQv.png", alt="Permission prompt with the new one-time option", width="800", height="560" %}

Reviewer: ```suggestion
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/7aSSSclHdEj33k5ifeQv.png", alt="Permission prompt with the new one-time option.", width="800", height="560" %}
```

  <figcaption>
    One-time permissions can be managed like other permissions, via the <strong>Site controls</strong> or <strong>Site settings</strong>.
  </figcaption>
</figure>

### User benefits and UX design of "Allow this time"

**Allow this time** provides users with:

- **Increased control.** Chrome user research has shown that users often are not ready to make a persistent decision for website permissions, trying to assess if they get value in return for granting permissions. Offering a one-time Allow option accommodates this need and makes allowing access less risky.
- **More clarity.** Explicit **Allow this time** and **Allow on every visit** options make it clear that one of the options is temporary, while the other one is persistent.

{% Aside %}
Chrome UX teams tested multiple layout and string variations before settling on the vertical three-button layout. User feedback indicated that a layout similar to mobile UIs provides safer outcomes and better expectation match. We think that this is partly due to users being already familiar with this pattern.

<figure>

> {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/wthfyv9PF4DUz5XnBbgk.jpg", alt="Permission prompt for an Android app", width="336", height="674" %}

Reviewer: ```suggestion
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/wthfyv9PF4DUz5XnBbgk.jpg", alt="Permission prompt for an Android app.", width="336", height="674" %}
```

For example:
* In Safari 16 on desktop, geolocation is one-time by default (until next navigation). The user can opt into access being persisted for 24 hours via a checkbox.
* In Firefox 115 on desktop, geolocation, camera and microphone permissions are one-time by default. The user can opt into persistent access via a checkbox.

{% Aside %}
One-time permissions are also built into several mobile operating systems.
{% endAside %}

## Impact and recommendations

Impact of one-time permissions:
- The introduction of one-time permissions does not impact permissions that users have previously granted persistently. 
- Once a user makes a decision on a permission prompt, then for the duration of their visit, the permission state is either granted or denied. One-time permissions make no difference in this regard. 
- However, if the user selects **Allow this time**, they will see a permission prompt again on their next visit. 

To accommodate for one-time permissions, we recommend that site owners do the following:

> - **Ensure your implementation follows best practices around permissions.** Review [Best practices](#best-practices).

Reviewer: ```suggestion
- Ensure your implementation follows best practices around permissions. Review [Best practices](#best-practices).
```

> - **Understand when one-time permissions expire.** See details in [Expiration of one-time permissions](#expiration).

Reviewer: ```suggestion
- Understand when one-time permissions expire. See details in [Expiration of one-time permissions](#expiration).
```

> - **Understand how the [Permissions API](https://developer.mozilla.org/docs/Web/API/Permissions_API) behaves** for one-time permissions, and **how to observe expiration.** Review details in [Permissions API](#permissions-api).

Reviewer: ```suggestion
- Understand how the [Permissions API](https://developer.mozilla.org/docs/Web/API/Permissions_API) behaves for one-time permissions, and **how to observe expiration.** Review details in [Permissions API](#permissions-api).
```

- Provide the necessary context before requesting permissions. Explain to your users why they'll benefit from a capability.
- Requests permissions at an appropriate and expected moment. Ideally, let your users express their intention to use a certain capability at their own pace.
- For the duration of the user’s visit, provide the same experience regardless of whether the user selected the one-time or persistent option.

Review additional guidance in [Permission UX](https://developers.google.com/web/fundamentals/push-notifications/permission-ux).

### Expiration of one-time permissions {: #expiration }

By default, all web permissions are bound to an [origin](https://web.dev/same-site-same-origin/#origin), which can be thought of as an app of its own. With **Allow this time**, the user gives a one-time permission to an origin. This permission lasts until the user stops actively interacting with that origin. In essence, one-time permissions expire when the user stops actively interacting with that origin for some time. This means the site can continue to use the capability if the user briefly switches to another web page in a different tab but then comes back.

> In more detail, one-time permission grants expire as soon as **any** of the following conditions are met:

Reviewer: ```suggestion
In more detail, one-time permission grants expire as soon as any of the following conditions are met:
```

* The page has been closed, was navigated away from, or was [discarded](/blog/page-lifecycle-api/#states:~:text=next%20states%3A%0ANONE-,Discarded,-A%20page%20is). This includes closing Chrome.
* 16 hours have passed since granting permission.

> * The user manually revokes the permission (e.g. in **Site controls**), or the permission is overridden through an enterprise policy.

Reviewer: ```suggestion
* The user manually revokes the permission (for example, in **Site controls**), or the permission is overridden through an enterprise policy.
```

> * The page has been in the [background](#background-tab) for at least 5 minutes — except if the capability is allowed to run in the background, like camera or microphone. In this case, as long as the site uses the capability, Chrome shows a [tab strip indicator](#tab-strip-indicator) and doesn't start the 5-minute timer until the page stops using the capability. Note that the 16-hour timer is still running.

Reviewer: ```suggestion
* The page has been in the [background](#background-tab) for at least 5 minutes—except if the capability is allowed to run in the background, like camera or microphone. In this case, as long as the site uses the capability, Chrome shows a [tab strip indicator](#tab-strip-indicator) and doesn't start the 5-minute timer until the page stops using the capability. Note that the 16-hour timer is still running.
```

{% Aside %}

> A user may have several tabs of the same origin open simultaneously. In that case, the expiration conditions are unchanged — you only need to replace the term "page" with "all pages from that origin".

Reviewer: ```suggestion
A user may have several tabs of the same origin open simultaneously. In that case, the expiration conditions are unchanged—you only need to replace the term "page" with "all pages from that origin".
```

{% endAside %}

#### Background tab {: #background-tab }

> A browser tab is in the *background* when it's not in the foreground. A tab is in the foreground when it's the visible tab in a browser window that is not minimized. 

Reviewer: ```suggestion
A browser tab is in the _background_ when it's not in the foreground. A tab is in the foreground when it's the visible tab in a browser window that is not minimized. 
```

> In Chrome, some capabilities, like **geolocation**, can only run in the **foreground tab**. They're not allowed to run in a background tab. Others, like **camera** and **microphone**, are allowed to run in a **background tab**. 

Reviewer: ```suggestion
In Chrome, some capabilities, like geolocation, can only run in the foreground tab. They're not allowed to run in a background tab. Others, like camera and microphone, are allowed to run in a background tab. 
```

> {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/Cjc8Tj8lnks8eB6Ntx7k.jpg", alt="ALT_TEXT_HERE", width="800", height="579" %}

Reviewer: ```suggestion
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/Cjc8Tj8lnks8eB6Ntx7k.jpg", alt="A screenshot of the browser window highlight an active foreground tab and inactive background tab.", width="800", height="579" %}
```

#### Tab strip indicator {: #tab-strip-indicator }

Capabilities that continue to run when the tab is in the background have a *tab strip indicator*. 
For example, camera has a tab strip indicator.

<figure>

> {% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/RE7sxBJLNpuvD7wUcOJW.jpg", alt="ALT_TEXT_HERE", width="515", height="111" %}

Reviewer: ```suggestion
{% Img src="image/O2RNUyVSLubjvENAT3e7JSdqSOx1/RE7sxBJLNpuvD7wUcOJW.jpg", alt="A tab with a red dot.", width="515", height="111" %}
```

### Permissions API {: #permissions-api }

{% Aside 'caution' %}

At the time of this writing, the [Permissions API](https://developer.mozilla.org/docs/Web/API/Permissions_API) is only available for some capabilities, depending on the browser. Review the list of capabilities that are permissions-aware [here](https://developer.mozilla.org/docs/Web/API/Permissions_API#permission-aware_apis), and detailed browser support [here](https://developer.mozilla.org/docs/Web/API/Permissions_API#browser_compatibility).
{% BrowserCompat 'api.Permissions' %}

{% endAside %}

To query the status of API permissions, you can use the [Permissions API](https://developer.mozilla.org/docs/Web/API/Permissions_API):

> - If the user picks **Allow this time**, the Permission API status is set to `granted`.  This means that an unexpired **one-time** permission and a **persistent** permission are indistinguishable and have the same status: `granted`.

Reviewer: ```suggestion
- If the user picks **Allow this time**, the Permission API status is set to `granted`.  This means that an unexpired one-time permission and a persistent permission are indistinguishable and have the same status: `granted`.
```
